### ISSUE TITLE
Autosave failure

### ISSUE THREAD

**Issue by jgsw** [Zettlr/Zettlr#4160]
### Description

I was working on a file for about an hour this morning, when I noticed that the file on disk had not been updated. This is despite the fact that autosave immediately was turned on. Attempting to manually save made no difference - Zettlr still failed to save. There was no indication from within the app that saving had failed, and no user-visible error message.  

### Reproducing

Something similar happened last week too, but unfortunately, I have no idea how to reproduce - it's something that happens rarely but without an obvious trigger.  The drive has plenty of space, so it was not a question e.g. of a full disk. 

Looking at the log, the only errors or warnings that look potentially relevant from the time in question are the following:

[10:43:07] [Warning] [R] [Main Window (d0e0bf0e-0358-40e7-8792-4009655fb8d5)] Could not apply delta, either local or remote not available [object Proxy] undefined 44754afa-42aa-4462-ab23-3cc3bf4d6f3f (index.js:2)
[10:43:07] [Warning] [R] [Main Window (d0e0bf0e-0358-40e7-8792-4009655fb8d5)] Could not apply delta, either local or remote not available [object Proxy] undefined 44754afa-42aa-4462-ab23-3cc3bf4d6f3f (index.js:2)
[10:43:07] [Error] [R] [Main Window (d0e0bf0e-0358-40e7-8792-4009655fb8d5)] Could not find removed leaf 44754afa-42aa-4462-ab23-3cc3bf4d6f3f or its local parent. Recovering full state. (index.js:2)
...
[10:43:54] [Error] [R] [Main Window (d0e0bf0e-0358-40e7-8792-4009655fb8d5)] Error: Error invoking remote method 'documents-authority': RangeError: Applying change set to a document with the wrong length (index.js:2)

### Zettlr Version

Beta (if applicable)

### Specify version

3.0.0Beta-2

### Installation Method

From the Website or GitHub

### Your Platform

- [X] Windows
- [ ] macOS
- [ ] Linux

### Architecture

- [X] Intel 64bit
- [ ] ARM 64bit

### Operating System Version

Windows 10

**Comment by pagiraud** [comment]
Confirming this bug. Just happened to me with Linux build. One hour of work lost.

**Comment by nathanlesage** [comment]
Would you mind re-trying this with autosave turned both completely off and "after a short delay"? I strongly suspect that this may be a race condition, where the backend attempts to save the file while new updates flow in, which upsets the state.

If we can confirm that giving it a longer time before saving this problem does not occur, I know that I basically just have to increase the save interval. (Or, maybe, remove the "immediate" function altogether, since 5 seconds do not seem too bad either, does it?)

To be sure, I did not encounter this problem at all, but I work with autosave turned off, so in my case, any changes to the document are separate from it being saved to disk.

**Comment by jgsw** [comment]
OK - I'll try with autosave turned off, and see if the problem recurs. 

(You'll probably have thought of this already, but thinking out loud: if the problem is caused by a race condition, it would be weird if this race condition occurred every single time Zettlr attempted to autosave the file over a period of an hour. So, assuming that the root cause is a race condition, then when this race condition occurs, it must somehow put the file into an unsaveable state.)

**Comment by nathanlesage** [comment]
> when this race condition occurs, it must somehow put the file into an unsaveable state

That's my gut feeling as well. The problem is that the documents we store are actually quite complex. When we update them, I don't really know what is happening (since we're using the component offered by CodeMirror to handle updates), and when we save them, we serialize them to a string. It could be if both actions occurred at the same time on the same object, that this puts the document itself into an unsaveable state.

I'll have a quick look and maybe get CodeMirror involved.

**Comment by nathanlesage** [comment]
Just checked the code for the document text container and … well, yes, it's pretty complex. I can imagine that simultaneously updating and serializing it to string could cause problems. Therefore, let me reiterate: try to reproduce this bug with autosave turned off. If that then does not cause this bug anymore, this means that I'll have to guard both the save and the update procedures so that they are mutually exclusive; i.e., saving will be delayed if the document is currently being updated, and vice versa.

**Comment by jgsw** [comment]
I've been using Zettlr for the past week with autosave turned off, and have not experienced any problems. So turning off autosave entirely seems to have fixed this issue for me. I remembered a couple of things that may be relevant in tracking down the bug. 

First, when the bug did occur, it also led to data loss when switching between tabs. I remember losing work when CTRL-tabbing between two open tabs. I'd modified one file, then used CTRL-tab to move focus away to a second file, and then CTRL-tabbed to move focus back to the first file. All the modifications I'd made to the first file were gone. (This might give some clues about whether the bug occurs in serializing to string).

Second, I have a hunch that the bug may have been triggered by Zotero pushing an update to my citation database at the same time as Zettlr was trying to autosave. The bug did seem to coincide with a couple of occasions when I'd just updated my Zotero database.

**Comment by nathanlesage** [comment]
No, I neither think Zotero syncs are the problem, nor that there is any problem serializing to string. Rather, it seems to be related with #4168. I will take this into account when fixing the other issue.
